src/view/nova/lib/theme.ts

```diff
@@ -0,0 +1,394 @@
+  borderEndEndRadius: ['borderEndEndRadius'],
+  borderEndStartRadius: ['borderEndStartRadius'],
+  borderLeftColor: ['borderLeftColor'],
+  borderRadius: ['borderRadius'],
+  borderRightColor: ['borderRightColor'],
+  borderStartColor: ['borderStartColor'],
+  borderStartEndRadius: ['borderStartEndRadius'],
+  borderStartStartRadius: ['borderStartStartRadius'],
+  borderStyle: ['borderStyle'],
+  borderTopColor: ['borderTopColor'],
+  borderTopEndRadius: ['borderTopEndRadius'],
+  borderTopLeftRadius: ['borderTopLeftRadius'],
+  borderTopRightRadius: ['borderTopRightRadius'],
+  borderTopStartRadius: ['borderTopStartRadius'],
+  opacity: ['opacity'],
+  elevation: ['elevation'],
+  pointerEvents: ['pointerEvents'],
+
+  // TextStyleIOS
+  fontVariant: ['fontVariant'],
+  textDecorationColor: ['textDecorationColor'],
+  textDecorationStyle: ['textDecorationStyle'],
+  writingDirection: ['writingDirection'],
+
+  // TextStyleAndroid
+  textAlignVertical: ['textAlignVertical'],
+  verticalAlign: ['verticalAlign'],
+  includeFontPadding: ['includeFontPadding'],
+
+  // TextStyle
+  color: ['color'],
+  fontFamily: ['fontFamily'],
+  fontSize: ['fontSize'],
+  fontStyle: ['fontStyle'],
+  fontWeight: ['fontWeight'],
+  letterSpacing: ['letterSpacing'],
+  lineHeight: ['lineHeight'],
+  textAlign: ['textAlign'],
+  textDecorationLine: ['textDecorationLine'],
+  textShadowColor: ['textShadowColor'],
+  textShadowOffset: ['textShadowOffset'],
+  textTransform: ['textTransform'],
+
+  // ImageStyle
+  resizeMode: ['resizeMode'],
+  overlayColor: ['overlayColor'],
+  tintColor: ['tintColor'],
+  objectFit: ['objectFit'],
+}
+
+export const createTheme = <
+  T extends Tokens,
+  P extends Properties,
+  M extends Macros<T>,
+  B extends Breakpoints,
+>({
+  tokens,
+  properties: userProperties = {},
+  macros: userMacros = {},
+  breakpoints: userBreakpoints = {},
+}: {
+  tokens: T
+  properties?: Partial<P>
+  macros?: Partial<M>
+  breakpoints?: Partial<B>
+}): Theme<T, P, M, B> => {
+  const properties = Object.assign({}, propertyMapping, userProperties) as P
+  const macros = userMacros as M
+  const breakpoints = Object.entries(userBreakpoints)
+    .sort(([_a, a], [_b, b]) => {
+      return a - b
+    })
+    .reduce((breakpoints, [key, value]) => {
+      breakpoints[key as keyof B] = value
+      return breakpoints
+    }, {} as B)
+
+  type InnerTheme = Theme<
+    typeof tokens,
+    typeof properties,
+    typeof macros,
+    typeof breakpoints
+  >
+
+  function _applyStyleProperty(styles: StyleObject, prop: string, value: any) {
+    for (const _prop of properties[prop]) {
+      // @ts-ignore no index sig, it's fine
+      const t = specialTokenMapping[_prop]
+        ? // @ts-ignore no index sig, it's fine
+          tokens[specialTokenMapping[_prop]]
+        : tokens[_prop]
+      styles[_prop] = t?.[value] || value
+    }
+  }
+
+  const style: InnerTheme['style'] = (rawProps, activeBreakpoints = []) => {
```

Comment:
How often is this function called? Does it get called during rendering, or during module init?

Reply:
It's called within `useStyle`, `useStyles`, and `styled`, which are all now memoized. I added examples of those hooks to the Debug screen.